hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ApiStatusEnricher.java

```diff
@@ -31,49 +34,74 @@
-
-      // if application tracer doesn't send the status message, then, we'll use
-      // our default mapping
-      if (statusMessage == null) {
-        statusMessage = GrpcCodeMapper.getMessage(statusCode);
-      }
+      statusMessage = getGrpcStatusMessage(event, statusCode);
     } else if (Protocol.PROTOCOL_HTTP == protocol || Protocol.PROTOCOL_HTTPS == protocol) {
       status = HttpCodeMapper.getState(statusCode);
 
-      statusMessage = SpanAttributeUtils.getStringAttribute(
-          event,
-          RawSpanConstants.getValue(Http.HTTP_RESPONSE_STATUS_MESSAGE));
-      if (statusMessage == null) {
-        statusMessage = HttpCodeMapper.getMessage(statusCode);
-      }
-
+      statusMessage = getHttpStatusMessage(event, statusCode);
     }
     addEnrichedAttributeIfNotNull(event, EnrichedSpanConstants.getValue(Api.API_STATUS_CODE), statusCode);
     addEnrichedAttributeIfNotNull(event, EnrichedSpanConstants.getValue(Api.API_STATUS_MESSAGE), statusMessage);
     addEnrichedAttributeIfNotNull(event, EnrichedSpanConstants.getValue(Api.API_STATUS), status);
   }
 
-  public static String getStatusCode(Event event, Protocol protocol) {
+  private static String getStatusCode(Event event, Protocol protocol) {
     List<String> statusCodeKeys = Lists.newArrayList();
     if (Protocol.PROTOCOL_GRPC == protocol) {
+      if (event.getGrpc() != null && event.getGrpc().getResponse() != null) {
+        int statusCode = event.getGrpc().getResponse().getStatusCode();
+        // Checking for the default value for status code field
+        if (statusCode != -1) {
+          return Integer.toString(statusCode);
+        }
+      }
       statusCodeKeys.add(RawSpanConstants.getValue(CensusResponse.CENSUS_RESPONSE_STATUS_CODE));
       statusCodeKeys.add(RawSpanConstants.getValue(Grpc.GRPC_STATUS_CODE));
       statusCodeKeys.add(RawSpanConstants.getValue(CensusResponse.CENSUS_RESPONSE_CENSUS_STATUS_CODE));
-
     } else if (Protocol.PROTOCOL_HTTP == protocol || Protocol.PROTOCOL_HTTPS == protocol) {
       statusCodeKeys.add(RawSpanConstants.getValue(OTSpanTag.OT_SPAN_TAG_HTTP_STATUS_CODE));
       statusCodeKeys.add(RawSpanConstants.getValue(Http.HTTP_RESPONSE_STATUS_CODE));
     }
 
     return SpanAttributeUtils.getFirstAvailableStringAttribute(event, statusCodeKeys);
   }
+
+  private static String getGrpcStatusMessage(Event event, String statusCode) {
+    String statusMessage = null;
+    if (event.getGrpc() != null && event.getGrpc().getResponse() != null) {
+      Response response = event.getGrpc().getResponse();
+      if (StringUtils.isNotBlank(response.getStatusMessage())) {
+        statusMessage = response.getStatusMessage();
+      } else if (StringUtils.isNotBlank(response.getErrorMessage())) {
+        statusMessage = response.getErrorMessage();
+      }
+    } else {
+      List<String> statusMessageKeys = Lists.newArrayList();
```

Comment:
Extract this as a constant since it's not changing from span to span.

Reply:
Done